File: felicia/examples/learn/message_communication/cc/node_create_flag.cc
#include "felicia/examples/learn/message_communication/cc/node_create_flag.h"

#include <iostream>

#include "felicia/core/util/command_line_interface/text_style.h"

namespace felicia {

NodeCreateFlag::NodeCreateFlag() {
  {
    BoolFlag::Builder builder(MakeValueStore(&is_publishing_node_));
    auto flag = builder.SetShortName("-p")
                    .SetHelp(
                        "create publishing node, if set, default: false(create "
                        "subscribing node)")
                    .Build();
    is_publishing_node_flag_ = std::make_unique<BoolFlag>(flag);
  }
  {
    StringFlag::Builder builder(MakeValueStore(&name_));
    auto flag = builder.SetLongName("--name").SetHelp("name for node").Build();
    name_flag_ = std::make_unique<StringFlag>(flag);
  }
  {
    StringFlag::Builder builder(MakeValueStore(&topic_));
    auto flag =
        builder.SetShortName("-t")
            .SetLongName("--topic")
            .SetHelp("topic to publish or subscribe, it's required to be set")
            .Build();
    topic_flag_ = std::make_unique<StringFlag>(flag);
  }
  {
    StringChoicesFlag::Builder builder(MakeValueStore<std::string>(
        &channel_type_, "TCP", Choices<std::string>{"TCP", "UDP"}));
    auto flag = builder.SetLongName("--channel_type")
                    .SetHelp(
                        "protocol to deliver message, it only works for "
                        "publishing node")
                    .Build();
    channel_type_flag_ = std::make_unique<StringChoicesFlag>(flag);
  }
}

NodeCreateFlag::~NodeCreateFlag() = default;

bool NodeCreateFlag::Parse(FlagParser& parser) {
  return PARSE_OPTIONAL_FLAG(parser, is_publishing_node_flag_, name_flag_,
                             topic_flag_, channel_type_flag_);
}

bool NodeCreateFlag::Validate() const {
  bool is_set = topic_flag_->is_set();
  if (!is_set) {
    std::cerr << TextStyle::Red("Error:") << " topic is not set." << std::endl;
  }
  return is_set;
}

}  // namespace felicia

File: felicia/examples/learn/message_communication/python/node_create_flag.py
from felicia.python.command_line_interface.flag_parser_delegate import FlagParserDelegate
import felicia_py.command_line_interface as cli


class NodeCreateFlag(FlagParserDelegate):
    def __init__(self):
        super().__init__()
        self.is_publshing_node_flag = cli.BoolFlagBuilder().set_short_name("-p").set_help(
            "create publishing node, if set, default: false(create subscribing node)").build()
        self.name_flag = cli.StringFlagBuilder().set_long_name(
            "--name").set_help("name for node").build()
        self.topic_flag = cli.StringFlagBuilder().set_short_name("-t").set_long_name(
            "--topic").set_help("topic to publish or subscribe, it's required to be set").build()
        self.channel_type_flag = cli.StringChoicesFlagBuilder("TCP", ["TCP", "UDP"]).set_long_name(
            "--channel_type").set_help(
                "protocol to deliver message, it only works for publishing node").build()

    def parse(self, flag_parser):
        return self.parse_optional_flags(flag_parser)

    def validate(self):
        is_set = self.topic_flag.is_set()
        if not is_set:
            print("{} tpoic is not set.".format(cli.TextStyle.red("Error:")))
        return is_set
